File: Difficulty: Medium/Geek's Training/geeks-training.cpp

#define vi vector<int>
#define vvi vector<vi>

class Solution {
  public:
    int recurSol(vvi& nums, int days, int task) {
        if(days<0) return 0;
        int maxi = INT_MIN;
        for(int i = 0;i<3;i++) {
            if(i==task) continue;
            int x = nums[days][i]+recurSol(nums,days-1,i);
            maxi = max(maxi,x);
        }
        return maxi;
    }
    
    int memoSol(vvi& nums, int days, int task,vvi& dp) {
        if(days<0) return 0;
        if(dp[days][task]!=-1) return dp[days][task];
        int maxi = INT_MIN;
        for(int i = 0;i<3;i++) {
            if(i==task) continue;
            int x = nums[days][i]+memoSol(nums,days-1,i,dp);
            maxi = max(maxi,x);
        }
        return dp[days][task] = maxi;
    }
    
    int maximumPoints(vector<vector<int>>& arr) {
        // Code here
        int days = arr.size();
        int index = days-1;
        // return recurSol(arr,days-1,3);
        vvi dp(days,vi(4,-1));
        return memoSol(arr,days-1,3,dp);
    }
};

File: Difficulty: Medium/Geek's Training/geeks-training.java
class Solution {

    public int solve(int[][] mat, int n, int m, int indexDay,int indexTask) {
        if(indexDay<0) return 0;
        
        int ans =Integer.MIN_VALUE;
        
        for(int i = 0;i<m;i++) {
            if(i==indexTask) continue;
            int x = solve(mat,n,m,indexDay-1,i) + mat[indexDay][i];
            ans = Math.max(ans,x);
        }
        
        return ans;
    }
    
    public int solve2(int[][] mat, int indexDay,int indexTask,int[][] dp) {
        if(indexDay<0) return 0;
        
        if(dp[indexDay][indexTask]!=-1) return dp[indexDay][indexTask];
        
        int ans =Integer.MIN_VALUE;
        
        for(int i = 0;i<3;i++) {
            if(i==indexTask) continue;
            int x = mat[indexDay][i] + solve2(mat,indexDay-1,i,dp);
            ans = Math.max(ans,x);
        }
        
        dp[indexDay][indexTask] = ans;
        
        return ans;
    }
    
    public int maximumPoints(int mat[][]) {
        // code here
        int n = mat.length;
        int m = mat[0].length;
        
        int indexDay = n-1;
        int indexTask = 3;
        
        // int ans = solve(mat,n,m,indexDay,indexTask);
        
        int[][] dp = new int[n][m+1];
        
        for(int i= 0;i<n;i++) {
            Arrays.fill(dp[i],-1);
        }
        
        int ans = solve2(mat,indexDay,indexTask,dp);
        
        return ans;
    }
}